components/sync15/src/client/sync.rs

```diff
@@ -40,54 +40,72 @@ pub fn synchronize_with_clients_engine(
     }
+    interruptee.err_if_interrupted()?;
+    // We assume an "engine" manages exactly one "collection" with the engine's name.
+    match engine.get_collection_request(coll_state.last_modified)? {
+        None => {
+            log::info!("skipping incoming for {} - not needed.", collection);
+        }
+        Some(collection_request) => {
+            // Ideally we would "batch" incoming records (eg, fetch just 1000 at a time)
+            // and ask the engine to "stage" them as they come in - but currently we just read
+            // them all in one request.
 
-    let collection_requests = engine.get_collection_requests(coll_state.last_modified)?;
-    let incoming = if collection_requests.is_empty() {
-        log::info!("skipping incoming for {} - not needed.", collection);
-        vec![IncomingChangeset::new(collection, coll_state.last_modified)]
-    } else {
-        assert_eq!(collection_requests.last().unwrap().collection, collection);
-
-        let count = collection_requests.len();
-        collection_requests
-            .into_iter()
-            .enumerate()
-            .map(|(idx, collection_request)| {
-                interruptee.err_if_interrupted()?;
-                let incoming_changes =
-                    super::fetch_incoming(client, &coll_state, collection_request)?;
-
-                log::info!(
-                    "Downloaded {} remote changes (request {} of {})",
-                    incoming_changes.changes.len(),
-                    idx,
-                    count,
-                );
-                Ok(incoming_changes)
-            })
-            .collect::<Result<Vec<_>, Error>>()?
+            // Doing this batching will involve specifying a "limit=" param and
+            // "x-if-unmodified-since" for each request, looking for an
+            // "X-Weave-Next-Offset header in the response and using that in subsequent
+            // requests.
+            // See https://mozilla-services.readthedocs.io/en/latest/storage/apis-1.5.html#syncstorage-paging
+            //
+            // But even if we had that, we need to deal with a 412 response on a subsequent batch,
+            // so we can't know if we've staged *every* record for that timestamp; the next
+            // sync must use an earlier one.
+            //
+            // For this reason, an engine can't really trust a server timestamp until the
+            // very end when we know we've staged them all.
+            let incoming = super::fetch_incoming(client, &coll_state, collection_request)?;
+            log::info!("Downloaded {} remote changes", incoming.len());
+            engine.stage_incoming(incoming, telem_engine)?;
+            interruptee.err_if_interrupted()?;
+        }
     };
 
-    let outgoing = engine.apply_incoming(incoming, telem_engine)?;
+    // Should probably add a new `fetch_outgoing()` and only call `apply()` when something was
+    // staged?
```

Comment:
Hmmm, what's this comment referring to? Do you mean skip calling `apply()` if we haven't staged any incoming changes?

Reply:
> My thinking is, if the engine stages incoming records in a temporary table, and is interrupted in the middle of apply

In that scenario I'd expect the engine should drop and create the temp tables as part of each sync.

> or if apply ends up "fixing up" any local items that weren't necessarily part of the last incoming batch

Do you mind expanding on this? You mean an engine might want to fixup a local item we've already uploaded in the past, which isn't currently marked as outgoing and wasn't incoming in this sync? I'm struggling to think where that might happen, but might be missing your point entirely.

Regardless, I updated that comment to point at this discussion :)